scraper/scraper.ts

```diff
@@ -1,28 +1,974 @@
-import * as puppeteer from 'puppeteer';
+import * as puppeteer from 'puppeteer'
+import {
+  Course,
+  CourseHead,
+  CourseInfo,
+  Time,
+  Class,
+  Chunk,
+  PageData,
+  TimetableData,
+  TimetableUrl,
+  UrlList,
+  Term,
+  ClassesByTerm,
+  ExtendedTerm,
+  Status,
+  Day,
+  Career,
+  classWarnings,
+  warning,
+} from './interfaces'
+import { cloneDeep } from 'lodash'
 
-// import * as .. does not work for chalk, so require.
-const chalk = require('chalk');
+/**
+ * Remove any html character entities from the given string
+ * At this point, it only looks for 3 of them as more are not necessary
+ * @param string The string to remove html characters from
+ */
+const removeHtmlSpecials = (string: string) => {
+  // &amp --> and
+  let newstr = string.replace('&amp;', 'and')
 
-const error = chalk.bold.red;
-const success = chalk.green;
-(async () => {
+  // &nbsp ---> nothing (as it appears in course enrolment when the course does not have one)
+  newstr = newstr.replace('&nbsp;', '')
+
+  // &lt --> < (less than), this could be changed to before??
+  newstr = newstr.replace('&lt;', '<')
+
+  // There was no greater than sign found, but if neccessary, can be added here
+
+  return newstr
+}
+
+/**
+ * Gets all the urls in the data class on page: page,
+ * given regex: regex.
+ * Each url will have the prefix: base.
+ * @param page Page to scrape urls from
+ * @param base base each url has to be prefixed with
+ * @param regex regex to check each url
+ */
+const getDataUrls = async (
+  page: puppeteer.Page,
+  base: string,
+  regex: RegExp
+): Promise<string[]> => {
+  // Get all the required urls...
+  const urls = await page.$$eval('.data', e => {
+    let inner = e.map(f => f.innerHTML)
+    return inner
+  })
```

Comment:
Can replace all the lines after this by building the urlSet as we go through the page
```suggestion
  const urls = await page.$$eval('a', (e: HTMLAnchorElement[]) => {
    const inner = e.map(f => f.href)
    return [...new Set(inner)]
  })
  
  return urls
```

Reply:
It won't let me convert it to an HTMLAnchorElement

The error:

No overload matches this call.
  Overload 1 of 5, '(selector: string, pageFunction: (elements: Element[]) => any): Promise<any>', gave the following error.
    Argument of type '(e: HTMLAnchorElement) => any' is not assignable to parameter of type '(elements: Element[]) => any'.
      Types of parameters 'e' and 'elements' are incompatible.
        Type 'Element[]' is missing the following properties from type 'HTMLAnchorElement': charset, coords, download, hreflang, and 262 more.
  Overload 2 of 5, '(selector: string, pageFunction: (elements: Element[], ...args: any[]) => any, ...args: SerializableOrJSHandle[]): Promise<any>', gave the following error.